ocaml/xapi/helpers.ml

```diff
@@ -1804,5 +1782,83 @@ end = struct
       raise e
 end
 
-let is_pool_secret_valid pool_secret =
-  List.exists (SecretString.equal pool_secret) !Xapi_globs.pool_secrets
+module PoolSecret : sig
+  val make : unit -> SecretString.t
+
+  val is_authorized : SecretString.t -> bool
+
+  val refresh_cache_or_create_new : unit -> unit
+end = struct
+  let to_secret x =
+    (* an opaque failure arises when a pool secret
+       contains a newline char, because a newline is
+       not encoded properly when putting the pool secret
+       into an http cookie. we enforce that the pool
+       secret has a certain form to avoid this kind
+       of issue
+    *)
+    let has_valid_chars =
+      Astring.(
+        String.for_all
+          (fun c -> Char.Ascii.is_alphanum c || c = '-' || c = '/')
+          x)
+    in
+    let sufficiently_secret = String.length x > 36 in
+    if has_valid_chars && sufficiently_secret |> not then
+      raise
+        Api_errors.(
+          Server_error
+            ( internal_error
+            , [
+                {|expected pool secret to match the following regex '^[0-9a-f\/\-]{37,}$'|}
+              ] )) ;
+    SecretString.of_string x
+
+  let _make () =
+    (* by default we generate the pool secret using /dev/urandom,
+       but if a script to generate the pool secret exists, use that instead *)
+    let make_urandom () =
+      Stdlib.List.init 3 (fun _ -> Uuid.(make_uuid_urnd () |> to_string))
+      |> String.concat "/"
+    in
+    let make_script () =
+      try call_script !Xapi_globs.gen_pool_secret_script []
+      with _ ->
+        info "helpers.ml:PoolSecret._make: script failed, using urandom instead" ;
```

Comment:
I've updated this so that if the script fails, we log and revert to /dev/urandom